pkg/controller/endpoint/endpoints_controller.go

```diff
@@ -184,18 +188,33 @@ func (e *EndpointController) Run(workers int, stopCh <-chan struct{}) {
 
 func (e *EndpointController) getPodServiceMemberships(pod *v1.Pod) (sets.String, error) {
 	set := sets.String{}
-	services, err := e.serviceLister.GetPodServices(pod)
+	allServices, err := e.serviceLister.Services(pod.Namespace).List(labels.Everything())
 	if err != nil {
 		// don't log this error because this function makes pointless
 		// errors when no services match.
 		return set, nil
 	}
-	for i := range services {
-		key, err := controller.KeyFunc(services[i])
-		if err != nil {
-			return nil, err
+
+	for i := range allServices {
+		service := allServices[i]
+		selector, ok := serviceSelectorCache[service]
```

Comment:
This isn't thread-safe and should be - there may be multiple workers accessing it the same time.

Reply:
Updated PR with usage of sync.Map